packages/platforms/accton/x86-64/as9726_32d/modules/builds/x86-64-accton-as9726-32d-psu.c

```diff
@@ -0,0 +1,266 @@
+	u8 status = 0;
+
+	mutex_lock(&data->update_lock);
+
+	data = as9726_32d_psu_update_device(dev);
+	if (!data->valid) {
+		mutex_unlock(&data->update_lock);
+		return sprintf(buf, "0\n");
+	}
+
+	if (attr->index == PSU_PRESENT) {
+		status = !(data->status >> (1-data->index) & 0x1);
+	} else { /* PSU_POWER_GOOD */
+		status = (data->status >> (3-data->index) & 0x1);
+	}
+
+	mutex_unlock(&data->update_lock);
+	return sprintf(buf, "%d\n", status);
+}
+
+static const struct attribute_group as9726_32d_psu_group = {
+	.attrs = as9726_32d_psu_attributes,
+};
+
+static int as9726_32d_psu_probe(struct i2c_client *client,
+				  const struct i2c_device_id *dev_id)
+{
+	struct as9726_32d_psu_data *data;
+	int status;
+
+	if (!i2c_check_functionality(client->adapter, 
+	    I2C_FUNC_SMBUS_I2C_BLOCK)) {
+		status = -EIO;
+		goto exit;
+	}
+
+	data = kzalloc(sizeof(struct as9726_32d_psu_data), GFP_KERNEL);
+	if (!data) {
+		status = -ENOMEM;
+		goto exit;
+	}
+
+	i2c_set_clientdata(client, data);
+	data->valid = 0;
+	data->index = dev_id->driver_data;
+	mutex_init(&data->update_lock);
+
+	dev_info(&client->dev, "chip found\n");
+
+	/* Register sysfs hooks */
+	status = sysfs_create_group(&client->dev.kobj, &as9726_32d_psu_group);
+	if (status)
+		goto exit_free;
+
+	data->hwmon_dev = hwmon_device_register_with_info(&client->dev, 
+							  "as9726_32d_psu", 
+							  NULL, NULL, NULL);
+	if (IS_ERR(data->hwmon_dev)) {
+		status = PTR_ERR(data->hwmon_dev);
+		goto exit_remove;
+	}
+
+	dev_info(&client->dev, "%s: psu '%s'\n",
+		dev_name(data->hwmon_dev), client->name);
+
+	return 0;
+
+exit_remove:
+	sysfs_remove_group(&client->dev.kobj, &as9726_32d_psu_group);
+exit_free:
+	kfree(data);
+exit:
+
+	return status;
+}
+
+static int as9726_32d_psu_remove(struct i2c_client *client)
+{
+	struct as9726_32d_psu_data *data = i2c_get_clientdata(client);
+
+	hwmon_device_unregister(data->hwmon_dev);
+	sysfs_remove_group(&client->dev.kobj, &as9726_32d_psu_group);
+	kfree(data);
+
+	return 0;
+}
+
+enum psu_index
+{
+	as9726_32d_psu1,
+	as9726_32d_psu2
+};
+
+static const struct i2c_device_id as9726_32d_psu_id[] = {
+	{ "as9726_32d_psu1", as9726_32d_psu1 },
+ 	{ "as9726_32d_psu2", as9726_32d_psu2 },
+	{}
+};
+MODULE_DEVICE_TABLE(i2c, as9726_32d_psu_id);
+
+static struct i2c_driver as9726_32d_psu_driver = {
+	.class = I2C_CLASS_HWMON,
+	.driver = {
+		.name = "as9726_32d_psu",
+	},
+	.probe = as9726_32d_psu_probe,
+	.remove = as9726_32d_psu_remove,
+	.id_table = as9726_32d_psu_id,
+	.address_list = normal_i2c,
+};
+
+static int as9726_32d_psu_read_block(struct i2c_client *client, u8 command,
+				      u8 *data, int data_len)
+{
+	int result = 0;
+	int retry_count = 5;
+
+	while (retry_count) {
+		
+		retry_count--;
+		result = i2c_smbus_read_i2c_block_data(client, command, 
+						       data_len, data);
+
+		if (unlikely(result < 0)) {
+			msleep(10);
+			continue;
+		}
+
+		if (unlikely(result != data_len)) {
+			result = -EIO;
+			msleep(10);
```

Comment:
Why such a high delay?

Reply:
10 ms delay is not so high here. I think.

Reply:
Everything about 1 ms is high, and should be justified.